### PR TITLE
HPCC-29523 Add sources to developer image

### PR DESCRIPTION
**Pull request by GordonSmith**
When building developer images using "image.sh", a snapshot of the sources should be included

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: dockerfiles/image.sh [hunk 1]
```diff
@@ -69,6 +69,25 @@ finalize_platform_core_image() {
...
+    if [ "$MODE" = "debug" ]; then
+        # Add sources
+        echo "--- Adding sources to '$image_name' image ---"
+        local CONTAINER=$(docker run -d \
+            --mount source=$ROOT_DIR,target=/hpcc-dev/HPCC-Platform-local,type=bind,readonly \
+            --mount source=$HPCC_BUILD,target=/hpcc-dev/build,type=volume \
+            hpccsystems/platform-core:$GIT_BRANCH-$MODE-$crc "tail -f /dev/null")
+        docker exec --user root --workdir /hpcc-dev/HPCC-Platform $CONTAINER /bin/bash -c "chown -R hpcc:hpcc /hpcc-dev/HPCC-Platform"
+        docker exec --workdir /hpcc-dev/HPCC-Platform $CONTAINER /bin/bash -c "git config --global --add safe.directory /hpcc-dev/HPCC-Platform-local"
+        docker exec --workdir /hpcc-dev/HPCC-Platform $CONTAINER /bin/bash -c "cp -r ../HPCC-Platform-local/.git ./.git"
```
Comment (jakesmith) [inline dockerfiles/image.sh:82]: minor: I know it's temporary, but .git (with history etc.) can be quite big/time consuming to copy.
I think you could instead copy everything with git archive like this:
```
cd ../HPCC-Platform-local
git archive --format=tar HEAD | tar xf - -C /hpcc-dev/HPCC-Platform
git submodule foreach --recursive 'git archive --format=tar HEAD | (cd /hpcc-dev/HPCC-Platform; mkdir -p $path; cd $path; tar xf -)'
```

**Review by jakesmith — APPROVED** [review]
@GordonSmith - looks good and works for me in testing.
approving, but 1 suggestion to potentially speed up the copy.

File: dockerfiles/image.sh [hunk 2]
```diff
@@ -69,6 +69,25 @@ finalize_platform_core_image() {
...
+        docker exec --workdir /hpcc-dev/HPCC-Platform $CONTAINER /bin/bash -c "git reset --hard --recurse-submodules"
+        docker exec --workdir /hpcc-dev/HPCC-Platform-local $CONTAINER /bin/bash -c "git ls-files --modified --exclude-standard -z | xargs -0 -I {} cp {} /hpcc-dev/HPCC-Platform/"
```
Comment (jakesmith) [inline dockerfiles/image.sh:84]: this needs to be:
... cp {} /hpcc-dev/HPCC-Platform/{}

(it's currently copying changed files in to the roo of /hpcc-dev/HPCC-Platform)

**Review by jakesmith — CHANGES_REQUESTED** [review]
@GordonSmith - noticed 1 issue in testing.

File: dockerfiles/image.sh [hunk 3]
```diff
@@ -44,11 +44,17 @@ create_build_image() {
...
     local base=$1
-    echo "--- Create 'platform-core:$MODE' image ---"
+    echo "--- Create 'platform-core:release' image ---"
     docker build --rm -f "$SCRIPT_DIR/vcpkg/platform-core-$BUILD_OS.dockerfile" \
-        -t platform-core:$MODE \
+        -t platform-core:release \
         --build-arg BASE_IMAGE=$base \
             "$SCRIPT_DIR/vcpkg/." 
+
+    echo "--- Create 'platform-core:debug' image ---"
```
Comment (jakesmith) [inline dockerfiles/image.sh:53]: question: now unconditionally building platform-core:release and platform-core:debug ?
Isn't that time consuming/wasteful?

**Review by jakesmith — COMMENTED** [review]
@GordonSmith - a couple of comments.

**Review by jakesmith — COMMENTED** [review]
as far as the image with sources/ability to debug looks good.

Might be unrelated, to these changes, building on my system, rebuilds and redownloads many packages (150+).

Discussing with @GordonSmith, it turned out to be related to the final step:
```
        finalize_platform_core_image $crc \
            "cmake --build /hpcc-dev/build --parallel --target install"
```

Gordon's suggested change of:

```
        finalize_platform_core_image $crc \
            "cmake --install /hpcc-dev/build --prefix /opt/HPCCSystems"
```

resolved the issue for me.

**Comment by GordonSmith** [comment]
@jakesmith I added that one line change.

**Review by jakesmith — APPROVED** [review]

**Review by jakesmith — APPROVED** [review]
@GordonSmith - looks good to me, please squash.

**Comment by GordonSmith** [comment]
@jakesmith @ghalliday squashed

**Pull request merged**